# Importing required modules and packages 

In [328]:
!conda install -c conda-forge geopy --yes; # uncomment this line if you haven't completed the Foursquare API lab

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.



In [329]:
!conda install -c conda-forge geocoder --yes;

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.



In [330]:
!conda install -c conda-forge folium=0.5.0 --yes;

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.



In [331]:
#from zipfile import ZipFile
from sklearn.cluster import KMeans # import k-means from clustering stage
import folium
import numpy as np  # useful for many scientific computing in Python
import pandas as pd # primary data structure library
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import requests
import io
import matplotlib.cm as cm
import matplotlib.colors as colors
import json

<div class="alert alert-block alert-danger">
<b> Skip the following lines until next red box:</b> Data is already downloaded, parsed and I already took a sample
</div>

## Downloading zip file and extracting it

In [17]:
#!wget -q -O 'zipsing' http://download.geonames.org/export/zip/SG.zip

In [18]:
#with ZipFile('zipsing', 'r') as zip:
#    zip.extract('SG.txt')

## Parsing the txt file and renaming columns / You can pass directly to the saved CSV file

In [89]:
data = pd.read_csv('SG.txt',sep='	', header=None)

In [326]:
data.head(3)

postalcode                  area  latitude  longitude
0       18906     Straits Boulevard    1.2758   103.8496
1       18907     Straits Boulevard    1.2749   103.8517
2       18910  Marina Gardens Drive    1.2796   103.8690

## Dropping and renaming columns

In [91]:
data.drop([0,3,4,5,6,7,8,11], axis = 1, inplace=True)
print(data.shape)

(121154, 4)


In [92]:
data.rename(columns={1:'postalcode',2:'area',9:'latitude',10:'longitude'}, inplace =True)

## Taking a sample of the original file, cleaning and sorting

The original file contains 121154 instances, this is why I will take a random sample

In [23]:
pcsubset = data.sample(n=320) #Postal code subset of the large database
pcsub = pcsubset[:] #Backing up the data

In [39]:
pcsub.to_csv('sampledata.csv') #Saving the dataframe as CSV in case the kernel dies

<div class="alert alert-block alert-danger">
<b> Use the csv stored file instead:</b> in the line below
</div>

In [334]:
pcsub = pd.read_csv('sampledata.csv') # Reading csv file stored locally

In [335]:
pcsub.drop('Unnamed: 0', axis = 1, inplace=True)

In [336]:
pcsub.sort_values(['area'], ascending=True, axis=0, inplace=True)

In [337]:
pcsub = pcsub.reset_index(drop=True)

In [338]:
pcsub.head()

postalcode                 area  latitude  longitude
0      677851        Almond Street    1.3725   103.7724
1       69878          Amoy Street    1.2801   103.8468
2      544616  Anchorvale Crescent    1.3985   103.8921
3      567920  Ang Mo Kio Avenue 2    1.3774   103.8350
4      561203  Ang Mo Kio Avenue 3    1.3676   103.8446

### Define map coordinates and loop to add markers from random postal codes

In [339]:
# Singapore latitude and longitude values
latitude = 1.290270
longitude = 103.8198

In [340]:
singmap = folium.Map(location=[latitude, longitude], zoom_start=12)

In [341]:
fgroup = folium.map.FeatureGroup() # instantiate a feature group for the incidents in the dataframe

In [342]:
for lat, lng, in zip(pcsub['latitude'], pcsub['longitude']):
    fgroup.add_child(
        folium.features.CircleMarker(
            [lat, lng],
            radius=1, # define how big you want the circle markers to be
            color="black",
            fill=True,
            fill_color="black",
            fill_opacity=0.5
        )
    )

# add incidents to map
singmap.add_child(fgroup)

In [343]:
print(pcsub.shape)
pcsub.head()

(320, 4)


postalcode                 area  latitude  longitude
0      677851        Almond Street    1.3725   103.7724
1       69878          Amoy Street    1.2801   103.8468
2      544616  Anchorvale Crescent    1.3985   103.8921
3      567920  Ang Mo Kio Avenue 2    1.3774   103.8350
4      561203  Ang Mo Kio Avenue 3    1.3676   103.8446

# Using Foursquare to get information from venues

<div class="alert alert-block alert-danger">
<b> Skip the following lines until next red box:</b> Foursquare data is already downloaded (Limited calls)
</div>

In [32]:
CLIENT_ID = 'HJLPKNFYSMPOPHISTM0LKEJHZC1MK2ENDRP1RWT5NAWSYMCN' # your Foursquare ID
CLIENT_SECRET = 'G1I1EM0WQWTCR0JH0H334RF5FJMZMTJGXSLICV01B2P2MMMU' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

In [33]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [36]:
snv = getNearbyVenues(names=pcsubset['area'], latitudes=pcsubset['latitude'], longitudes=pcsubset['longitude']) #Singapore nearby venues

Almond Street
Amoy Street
Anchorvale Crescent
Ang Mo Kio Avenue 2
Ang Mo Kio Avenue 3
Ang Mo Kio Avenue 3
Ang Mo Kio Avenue 3
Ash Grove
Ashwood Grove
Barker Road
Bedok Avenue
Bedok Reservoir Road
Bedok Reservoir Road
Bedok Ria Walk
Bedok Road
Belgravia Drive
Belgravia Drive
Bilal Lane
Blandford Drive
Bodmin Drive
Boon Lay Drive
Brookvale Drive
Bukit Merah Central
Burghley Drive
Cairnhill Road
Cambridge Road
Canberra Drive
Cardiff Grove
Cashew Terrace
Casuarina Road
Casuarina Road
Cavan Road
Chancery Hill Walk
Chancery Lane
Changi Road
Charlton Road
Cheng Soon Crescent
Chin Terrace
Chinese Cemetery Path 4
Chiselhurst Grove
Choa Chu Kang Avenue 5
Choa Chu Kang Loop
Choa Chu Kang North 6
Clover Way
Cluny Park
Cluny Road
Compassvale Drive
Coral Island
Corfe Place
Cornwall Gardens
Coronation Road
Corporation Walk
Cuff Road
Cypress Avenue
Da Silva Lane
Dairy Farm Crescent
Deal Road
Dedap Link
Dunbar Walk
Dunlop Street
Dunman Lane
Dunsfold Drive
Duxton Hill
East Coast Road
East Coast Road
Eas

In [37]:
print(snv.shape)
bnv = snv[:]
bnv.head(3)

(17525, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Almond Street                 1.3725                103.7724   
1  Almond Street                 1.3725                103.7724   
2  Almond Street                 1.3725                103.7724   

                    Venue  Venue Latitude  Venue Longitude   Venue Category  
0     正味 卤鸭面 Braised Duck        1.376867       103.773162     Noodle House  
1    Chestnut Nature Park        1.371641       103.779792  Nature Preserve  
2  Dairy Farm Nature Park        1.364448       103.776364  Nature Preserve

In [38]:
bnv.to_csv('foursquare.csv') #Saving the dataframe as CSV in case the kernel dies

<div class="alert alert-block alert-danger">
<b> Use the code below:</b> to bring the foursquare data back
</div>

In [344]:
bnv = pd.read_csv('foursquare.csv') # Reading csv file stored locally

## One hot encoding / Preparing data to use in the K-means clustering

In [345]:
ctgrtable = pd.get_dummies(bnv[['Venue Category']], prefix="", prefix_sep="") 
ctgrtable.shape

(17525, 344)

In [346]:
ctgrtable.head(5)

Accessories Store  American Restaurant  Aquarium  Arcade  \
0                  0                    0         0       0   
1                  0                    0         0       0   
2                  0                    0         0       0   
3                  0                    0         0       0   
4                  0                    0         0       0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  \
0                     0                 0                   0   
1                     0                 0                   0   
2                     0                 0                   0   
3                     0                 0                   0   
4                     0                 0                   0   

   Australian Restaurant  Auto Garage  Auto Workshop  Automotive Shop  \
0                      0            0              0                0   
1                      0            0              0                0   
2                      0            0              0                0   
3                      0            0              0                0   
4                      0            0              0                0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Beach  Beach Bar  Bed & Breakfast  \
0                 0                 0      0          0                0   
1                 0                 0      0          0                0   
2                 0                 0      0          0                0   
3                 0                 0      0          0                0   
4                 0                 0      0          0                0   

   Beer Bar  Beer Garden  Beer Store  Belgian Restaurant  Betting Shop  \
0         0            0           0                   0             0   
1         0            0           0                   0             0   
2         0            0           0                   0             0   
3         0            0           0                   0             0   
4         0            0           0                   0             0   

   Bike Rental / Bike Share  Bike Trail  Bistro  Board Shop  Boarding House  \
0                         0           0       0           0               0   
1                         0           0       0           0               0   
2                         0           0       0           0               0   
3                         0           0       0           0               0   
4                         0           0       0           0               0   

   Boat or Ferry  Bookstore  Botanical Garden  Boutique  Bowling Alley  \
0              0          0                 0         0              0   
1              0          0                 0         0              0   
2              0          0                 0         0              0   
3              0          0                 0         0              0   
4              0          0                 0         0              0   

   Bowling Green  Breakfast Spot  Brewery  Bridal Shop  Bridge  \
0   

In [347]:
ctgrtable.drop("Neighborhood", axis = 1, inplace=True) # Dropping a column called Neighborhood that came directly from the Foursquare database

In [348]:
temptb = pd.concat([bnv["Neighborhood"].reset_index(drop=True),ctgrtable.reset_index(drop=True)], axis=1)
temptb.head()

Neighborhood  Accessories Store  American Restaurant  Aquarium  Arcade  \
0  Almond Street                  0                    0         0       0   
1  Almond Street                  0                    0         0       0   
2  Almond Street                  0                    0         0       0   
3  Almond Street                  0                    0         0       0   
4  Almond Street                  0                    0         0       0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  \
0                     0                 0                   0   
1                     0                 0                   0   
2                     0                 0                   0   
3                     0                 0                   0   
4                     0                 0                   0   

   Australian Restaurant  Auto Garage  Auto Workshop  Automotive Shop  \
0                      0            0              0                0   
1                      0            0              0                0   
2                      0            0              0                0   
3                      0            0              0                0   
4                      0            0              0                0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Beach  Beach Bar  Bed & Breakfast  \
0                 0                 0      0          0                0   
1                 0                 0      0          0                0   
2                 0                 0      0          0                0   
3                 0                 0      0          0                0   
4                 0                 0      0          0                0   

   Beer Bar  Beer Garden  Beer Store  Belgian Restaurant  Betting Shop  \
0         0            0           0                   0             0   
1         0            0           0                   0             0   
2         0            0           0                   0             0   
3         0            0           0                   0             0   
4         0            0           0                   0             0   

   Bike Rental / Bike Share  Bike Trail  Bistro  Board Shop  Boarding House  \
0                         0           0       0           0               0   
1                         0           0       0           0               0   
2                         0           0       0           0               0   
3                         0           0       0           0               0   
4                         0           0       0           0               0   

   Boat or Ferry  Bookstore  Botanical Garden  Boutique  Bowling Alley  \
0              0          0                 0         0              0   
1              0          0                 0         0              0   
2              0          0                 0         0              0   
3              0          0                 0         0              0   
4              0          0                 0         0 

In [349]:
groupt = temptb.groupby('Neighborhood').sum().reset_index()
groupt.head()
print(groupt.shape)

(291, 344)


# The groupt dataframe will be used in a paralel way for generating the top venues list but also to create the final dataframe for the clustering

## 1.- Preparing the dataframe for the k-mean clustering

In [350]:
forcluster = groupt[:] # As a backup but also is the dataframe to be used for clustering purposes

In [351]:
forcluster['latitude'] = forcluster.Neighborhood.map(pcsub.set_index('area')['latitude'].to_dict()) # Bringing the original latitude column

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [352]:
forcluster['longitude'] = forcluster.Neighborhood.map(pcsub.set_index('area')['longitude'].to_dict()) # Bringing the original longitude column

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [353]:
print(forcluster.shape) # This dataframe includes also the latitude and longitude
forcluster.head(2)

(291, 346)


Neighborhood  Accessories Store  American Restaurant  Aquarium  Arcade  \
0  Almond Street                  0                    0         0       0   
1    Amoy Street                  0                    0         0       0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       1            1           0                    0   

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  \
0                     0                 0                   1   
1                     0                 1                   0   

   Australian Restaurant  Auto Garage  Auto Workshop  Automotive Shop  \
0                      0            0              0                0   
1                      0            0              0                0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       3     0    4               0   

   Baseball Stadium  Basketball Court  Beach  Beach Bar  Bed & Breakfast  \
0                 0                 1      0          0                0   
1                 0                 0      0          0                0   

   Beer Bar  Beer Garden  Beer Store  Belgian Restaurant  Betting Shop  \
0         0            0           0                   0             0   
1         1            2           0                   0             0   

   Bike Rental / Bike Share  Bike Trail  Bistro  Board Shop  Boarding House  \
0                         0           0       0           0               0   
1                         0           0       0           0               0   

   Boat or Ferry  Bookstore  Botanical Garden  Boutique  Bowling Alley  \
0              0          0                 0         0              0   
1              0          1                 0         0              0   

   Bowling Green  Breakfast Spot  Brewery  Bridal Shop  Bridge  \
0              0               0        0            0       0   
1              0               0        1            0       0   

   Bubble Tea Shop  Buddhist Temple  Buffet  Building  Burger Joint  \
0                0                0       0         0             0   
1                0                1       0         0             0   

   Burrito Place  Bus Line  Bus Station  Bus Stop  Butcher  Cafeteria  Café  \
0              0         0            1         0        0          1     0   
1              1         0            0         0        0          0     4   

   Campground  Canal  Canal Lock  Candy Store  Cantonese Restaurant  \
0           0      0           0            0                     0   
1           0      0           0            0                     0   

   Caribbean Restaurant  Casino  Cha Chaan Teng  Cheese Shop  \
0                     0       0               0            0   
1                     0       0               0            0   

   Chinese Aristocrat Restaurant  Chinese Breakfast Place  Chinese Restaurant  \
0                              0                        0                   0   
1                              0                        0                   2   

   Chocolate Shop  Churrascaria  Climbing Gym  Clothing Store  Club House  \
0               0             0             0               0           0   
1               0             0             0               0           0   

   Cocktail Bar  Coffee Shop  College Administrative Building  \
0             0            1                                0   
1             2            5                                0   

   College Cafeteria  College Gym  College Residence Hall  College Stadium  \
0                  0            0                       0                0   
1                  0            0                       0                0   

   College Theater  Com

## 2.- Top venues per neighborhood

In [354]:
num_top_venues = 10

for hood in groupt['Neighborhood']:
    print("----"+hood+"----")
    temp = groupt[groupt['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Almond Street----
                venue  freq
0     Nature Preserve   3.0
1                Park   3.0
2          Food Court   3.0
3        Noodle House   2.0
4              Market   1.0
5         Coffee Shop   1.0
6  Seafood Restaurant   1.0
7         Flea Market   1.0
8             Parking   1.0
9           Cafeteria   1.0


----Amoy Street----
                 venue  freq
0                Hotel   7.0
1  Japanese Restaurant   6.0
2          Coffee Shop   5.0
3    Korean Restaurant   5.0
4                  Bar   4.0
5           Food Court   4.0
6                 Café   4.0
7           Restaurant   3.0
8               Bakery   3.0
9             Wine Bar   3.0


----Anchorvale Crescent----
                  venue  freq
0  Fast Food Restaurant   4.0
1            Food Court   4.0
2        Cosmetics Shop   3.0
3           Supermarket   2.0
4     Convenience Store   2.0
5           Coffee Shop   2.0
6    Chinese Restaurant   1.0
7  Gym / Fitness Center   1.0
8                   Gym   1.0

In [355]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [356]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
topv = pd.DataFrame(columns=columns)
topv['Neighborhood'] = groupt['Neighborhood']

for ind in np.arange(groupt.shape[0]):
    topv.iloc[ind, 1:] = return_most_common_venues(groupt.iloc[ind, :], num_top_venues)

topv.shape


(291, 11)

In [357]:
topv.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0        Almond Street                  Park       Nature Preserve   
1          Amoy Street                 Hotel   Japanese Restaurant   
2  Anchorvale Crescent  Fast Food Restaurant            Food Court   
3  Ang Mo Kio Avenue 2    Chinese Restaurant            Food Court   
4  Ang Mo Kio Avenue 3            Food Court    Chinese Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Food Court          Noodle House    Seafood Restaurant   
1           Coffee Shop     Korean Restaurant                  Café   
2        Cosmetics Shop     Convenience Store           Supermarket   
3      Asian Restaurant             Pet Store   Japanese Restaurant   
4           Coffee Shop  Fast Food Restaurant    Seafood Restaurant   

  6th Most Common Venue          7th Most Common Venue 8th Most Common Venue  \
0      Basketball Court                         Market                 Trail   
1            Food Court                            Bar            Restaurant   
2           Coffee Shop                         Bistro      Basketball Court   
3  Fast Food Restaurant                Thai Restaurant            Restaurant   
4           Supermarket  Vegetarian / Vegan Restaurant          Dessert Shop   

  9th Most Common Venue 10th Most Common Venue  
0               Parking            Flea Market  
1                Bakery     Spanish Restaurant  
2                   Gym             Restaurant  
3           Coffee Shop                    Pub  
4          Noodle House    Japanese Restaurant

In [358]:
topa = topv[:] # Backup

In [359]:
topa.rename(columns={'Neighborhood':'area'}, inplace =True)
print(topa.shape)


(291, 11)


## Running K-Means using the first part

In [360]:
# set number of clusters
kclusters = 10

tgc = forcluster.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tgc)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([6, 3, 6, 9, 4, 2, 6, 2, 6, 0], dtype=int32)

In [361]:
topa.shape

(291, 11)

In [362]:
pcsub.shape

(320, 4)

Let's merge the dataframes that includes the cluster as well as the top 10 venues for each neighborhood.

In [363]:
# add clustering labels
topa.insert(0, 'Cluster Labels', kmeans.labels_)

tablemerged= pcsub[:]

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
tablemerged = tablemerged.join(topa.set_index('area'), on='area')

In [364]:
tablemerged = tablemerged.drop('postalcode',1)

In [365]:
tablemerged.head(5)

area  latitude  longitude  Cluster Labels  \
0        Almond Street    1.3725   103.7724               6   
1          Amoy Street    1.2801   103.8468               3   
2  Anchorvale Crescent    1.3985   103.8921               6   
3  Ang Mo Kio Avenue 2    1.3774   103.8350               9   
4  Ang Mo Kio Avenue 3    1.3676   103.8446               4   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                  Park       Nature Preserve            Food Court   
1                 Hotel   Japanese Restaurant           Coffee Shop   
2  Fast Food Restaurant            Food Court        Cosmetics Shop   
3    Chinese Restaurant            Food Court      Asian Restaurant   
4            Food Court    Chinese Restaurant           Coffee Shop   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0          Noodle House    Seafood Restaurant      Basketball Court   
1     Korean Restaurant                  Café            Food Court   
2     Convenience Store           Supermarket           Coffee Shop   
3             Pet Store   Japanese Restaurant  Fast Food Restaurant   
4  Fast Food Restaurant    Seafood Restaurant           Supermarket   

           7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                         Market                 Trail               Parking   
1                            Bar            Restaurant                Bakery   
2                         Bistro      Basketball Court                   Gym   
3                Thai Restaurant            Restaurant           Coffee Shop   
4  Vegetarian / Vegan Restaurant          Dessert Shop          Noodle House   

  10th Most Common Venue  
0            Flea Market  
1     Spanish Restaurant  
2             Restaurant  
3                    Pub  
4    Japanese Restaurant

In [391]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tablemerged['latitude'], tablemerged['longitude'], tablemerged['area'], tablemerged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color= 'black',
        weight=1,
        fill=True,
        fill_color= rainbow[cluster -1],
        fill_opacity=1).add_to(map_clusters)

    
    
    
    
map_clusters

### Cluster 0: Mixed restaurants and coffee shops in different areas

In [367]:
tablemerged.loc[tablemerged['Cluster Labels'] == 0, tablemerged.columns[[0] + list(range(4, tablemerged.shape[1]))]]

area 1st Most Common Venue 2nd Most Common Venue  \
11        Bedok Reservoir Road            Food Court           Coffee Shop   
12        Bedok Reservoir Road            Food Court           Coffee Shop   
20              Boon Lay Drive      Asian Restaurant  Fast Food Restaurant   
26              Canberra Drive           Coffee Shop  Fast Food Restaurant   
34                 Changi Road    Chinese Restaurant           Coffee Shop   
43                  Clover Way           Coffee Shop    Chinese Restaurant   
69             Enterprise Road   Japanese Restaurant      Asian Restaurant   
71                 Faber Green            Food Court           Coffee Shop   
74               Fernvale Road           Coffee Shop            Food Court   
85               Ghim Moh Link           Coffee Shop   Japanese Restaurant   
93            Guan Soon Avenue                  Café           Coffee Shop   
96               Harvey Avenue           Coffee Shop                  Café   
103         Holland Grove View    Chinese Restaurant           Coffee Shop   
106           Hougang Avenue 8           Coffee Shop            Food Court   
113             Jalan Binchang            Food Court           Coffee Shop   
120             Jalan Kemuning           Coffee Shop    Chinese Restaurant   
123    Jalan Lengkok Sembawang  Fast Food Restaurant            Food Court   
124          Jalan Limau Manis           Coffee Shop            Food Court   
129                Jalan Paras     Indian Restaurant            Food Court   
130            Jalan Pari Unak           Coffee Shop            Food Court   
133                Jalan Punai            Food Court           Coffee Shop   
135              Jalan Selamat     Indian Restaurant            Food Court   
139              Jalan Soo Bee           Coffee Shop    Chinese Restaurant   
140           Jalan Tanah Rata           Coffee Shop          Noodle House   
150      Jurong West Street 61  Fast Food Restaurant      Asian Restaurant   
161              Lengkong Tiga    Chinese Restaurant     Indian Restaurant   
168                 Limau Rise           Coffee Shop            Food Court   
170         Lorong 1 Toa Payoh    Chinese Restaurant           Flower Shop   
177       Lorong G Telok Kurau    Chinese Restaurant          Noodle House   
180             Lorong Marzuki           Coffee Shop    Chinese Restaurant   
218              Parry Terrace           Coffee Shop                  Café   
219         Pasir Panjang Road           Coffee Shop            Food Court   
220         Pasir Panjang Road           Coffee Shop            Food Court   
222        Pasir Ris Street 12           Coffee Shop           Bus Station   
226             Phoenix Garden           Coffee Shop  Fast Food Restaurant   
242            Sea Breeze Walk           Coffee Shop    Chinese Restaurant   
245                 Senja Road  Fast Food Restaurant           Coffee Shop   
248   Serangoon North Avenue 4            Food Court           Coffee Shop   
250              Serenade Walk            Food Court    Chinese Restaurant   
252               Siglap Plain           Coffee Shop    Chinese Restaurant   
257      South Woodlands Drive            Food Court                  Café   
275          Tampines Avenue 8           Coffee Shop  Fast Food Restaurant   
276          Tampines Avenue 8           Coffee Shop  Fast Food Restaurant   
280  Tanah Merah Kechil Avenue          Noodle House           Coffee Shop   
287        Thomson Hills Drive            Food Court    Chinese Restaurant   
298               Ubi Crescent      Asian Restaurant            Food Court   
309           West Coast Green            Food Court   Japanese Restaurant   
310           West Coast Grove            Food Court     Indian Restaurant   
316          Yio Chu Kang Road      Asian Restaurant    Chinese Restaurant   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11       Asian Restaurant  Fast Food Restaurant     

### Cluster 1: Traditional asian restaurants

In [368]:
tablemerged.loc[tablemerged['Cluster Labels'] == 1, tablemerged.columns[[0] + list(range(4, tablemerged.shape[1]))]]

area 1st Most Common Venue 2nd Most Common Venue  \
35          Charlton Road          Noodle House                  Café   
60            Dunman Lane    Chinese Restaurant          Noodle House   
63        East Coast Road    Chinese Restaurant           Coffee Shop   
64        East Coast Road    Chinese Restaurant           Coffee Shop   
70           Everitt Road      Asian Restaurant    Chinese Restaurant   
86           Goodman Road    Chinese Restaurant          Noodle House   
87           Goodman Road    Chinese Restaurant          Noodle House   
95       Haigsville Drive    Chinese Restaurant      Asian Restaurant   
109         Jalan Baiduri    Chinese Restaurant          Noodle House   
147      Joo Chiat Avenue    Chinese Restaurant          Noodle House   
148       Joo Chiat Place    Chinese Restaurant          Noodle House   
149       Joo Chiat Place    Chinese Restaurant          Noodle House   
172     Lorong 14 Geylang    Chinese Restaurant          Noodle House   
173     Lorong 28 Geylang    Chinese Restaurant          Noodle House   
174     Lorong 37 Geylang            Food Court    Chinese Restaurant   
178  Lorong J Telok Kurau    Chinese Restaurant      Asian Restaurant   
186        Lorong Stangee          Noodle House    Chinese Restaurant   
191         Marine Parade    Chinese Restaurant          Noodle House   
256           Sims Avenue    Chinese Restaurant          Noodle House   
285        Tembeling Road    Chinese Restaurant          Noodle House   

    3rd Most Common Venue 4th Most Common Venue  \
35     Chinese Restaurant           Coffee Shop   
60       Asian Restaurant                  Café   
63      Indian Restaurant          Noodle House   
64      Indian Restaurant          Noodle House   
70           Noodle House     Indian Restaurant   
86       Asian Restaurant    Italian Restaurant   
87       Asian Restaurant    Italian Restaurant   
95      Indian Restaurant          Noodle House   
109           Coffee Shop      Asian Restaurant   
147           Coffee Shop      Asian Restaurant   
148      Asian Restaurant           Coffee Shop   
149      Asian Restaurant           Coffee Shop   
172    Seafood Restaurant             BBQ Joint   
173            Food Court    Seafood Restaurant   
174          Noodle House      Asian Restaurant   
178          Noodle House           Coffee Shop   
186      Asian Restaurant     Indian Restaurant   
191      Asian Restaurant     Indian Restaurant   
256      Asian Restaurant            Food Court   
285      Asian Restaurant     Indian Restaurant   

             5th Most Common Venue 6th Most Common Venue  \
35                      Food Court      Asian Restaurant   
60               Indian Restaurant            Food Court   
63                Asian Restaurant                  Café   
64                Asian Restaurant                  Café   
70                     Coffee Shop                 Hotel   
86                            Café            Food Court   
87                            Café            Food Court   
95                      Food Court                 Hotel   
109                     Food Court                Bakery   
147                     Food Court     Indian Restaurant   
148              Indian Restaurant    Seafood Restaurant   
149              Indian Restaurant    Seafood Restaurant   
172                     Food Court                 Hotel   
173  Vegetarian / Vegan Restaurant                 Hotel   
174  Vegetarian / Vegan Restaurant  Fast Food Restaurant   
178              Indian Restaurant            Food Court   
186                          Hotel                Bakery   
191                          Hotel                Bakery   
256  Vegetarian / Vegan Restaurant    Seafood Restaurant   
285                          Hotel                Bakery   

    7th Most Common Venue          8th Most Common Venue  \
35           Dessert Shop                Thai Restaurant   
60                    Bar         

### Cluster 2: Bakeries and cafes

In [369]:
tablemerged.loc[tablemerged['Cluster Labels'] == 2, tablemerged.columns[[0] + list(range(4, tablemerged.shape[1]))]]

area         1st Most Common Venue  \
7                 Ash Grove                        Bakery   
9               Barker Road                          Café   
44               Cluny Park                          Café   
45               Cluny Road                          Café   
50          Coronation Road                          Café   
53           Cypress Avenue                   Bus Station   
73              Farrer Road                          Café   
81            Gardenia Road                          Café   
91         Greenwood Avenue            Italian Restaurant   
92         Greenwood Avenue            Italian Restaurant   
108            Jalan Bahasa                          Café   
136              Jalan Seni                          Café   
159           Larkhill Road  Theme Park Ride / Attraction   
229  Princess Of Wales Road                          Café   
284   Telok Blangah Heights                          Café   
292        Trevose Crescent           Japanese Restaurant   
293        Trevose Crescent           Japanese Restaurant   
297             Tudor Close                          Café   
301              Venus Road               Thai Restaurant   
305            Watten Close                          Café   
306             Watten Rise                          Café   
307          Watten Terrace                          Café   
308             Watten View           Japanese Restaurant   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
7      Italian Restaurant            Food Court           Pizza Place   
9     Japanese Restaurant    Chinese Restaurant           Bus Station   
44                 Garden                Bakery                  Park   
45                 Garden                Bakery                  Park   
50                 Bakery         Shopping Mall          Noodle House   
53             Food Court    Seafood Restaurant     Indian Restaurant   
73                 Garden                Bakery     French Restaurant   
81        Thai Restaurant        Ice Cream Shop    Chinese Restaurant   
91                 Bakery                  Pool                   Gym   
92                 Bakery                  Pool                   Gym   
108                Bakery          Noodle House         Shopping Mall   
136         Shopping Mall                Bakery   Japanese Restaurant   
159            Theme Park                 Hotel                  Café   
229                Bakery    Italian Restaurant         Shopping Mall   
284            Food Court                 Trail   Sporting Goods Shop   
292         Metro Station          Noodle House         Shopping Mall   
293         Metro Station          Noodle House         Shopping Mall   
297          Noodle House   Japanese Restaurant         Shopping Mall   
301                  Café    Chinese Restaurant        Ice Cream Shop   
305                Bakery   Japanese Restaurant    Italian Restaurant   
306                Bakery   Japanese Restaurant    Italian Restaurant   
307                Bakery   Japanese Restaurant    Italian Restaurant   
308                  Café    Italian Restaurant                Bakery   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
7      Seafood Restaurant     Indian Restaurant           Coffee Shop   
9      Italian Restaurant        Sandwich Place  Fast Food Restaurant   
44      French Restaurant            Food Court         Shopping Mall   
45               Bus Stop     French Restaurant         Shopping Mall   
50        Thai Restaurant           Coffee Shop    Chinese Restaurant   
53                 Bakery   Japanese Restaurant    Italian Restaurant   
73          Shopping Mall          Noodle House         Metro Station   
81                    Spa           Coffee Shop                 Trail   
91     Chinese Restaurant   Japanese Restaurant           Pizza Place   
92     Chinese Restaurant   Japanese Restaurant           Pizza Place   
10

### Cluster 3: Business amenities 

In [370]:
tablemerged.loc[tablemerged['Cluster Labels'] == 3, tablemerged.columns[[0] + list(range(4, tablemerged.shape[1]))]]

area 1st Most Common Venue 2nd Most Common Venue  \
1           Amoy Street                 Hotel   Japanese Restaurant   
32   Chancery Hill Walk                  Café    Chinese Restaurant   
33        Chancery Lane    Chinese Restaurant                  Café   
62          Duxton Hill   Japanese Restaurant           Coffee Shop   
82         Gentle Drive    Chinese Restaurant                 Hotel   
97        Havelock Road    Chinese Restaurant   Japanese Restaurant   
104        Holland Rise                Bakery   Japanese Restaurant   
128        Jalan Novena    Chinese Restaurant                 Hotel   
143    Jalan Tiga Ratus                  Café           Coffee Shop   
210         Nassim Road                 Hotel                  Café   
212   North Bridge Road                 Hotel       Thai Restaurant   
213   North Bridge Road                 Hotel       Thai Restaurant   
216       Pagoda Street   Japanese Restaurant                 Hotel   
231       Punggol Field    Chinese Restaurant  Fast Food Restaurant   
232       Punggol Field    Chinese Restaurant  Fast Food Restaurant   
233       Punggol Field    Chinese Restaurant  Fast Food Restaurant   
262        Sunbird Road                  Café           Coffee Shop   
272       Taman Permata    Chinese Restaurant                  Café   
273       Taman Permata    Chinese Restaurant                  Café   
286       Temple Street            Food Court   Japanese Restaurant   
291         Tras Street   Japanese Restaurant           Coffee Shop   
303     Victoria Street                 Hotel                Bakery   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1             Coffee Shop     Korean Restaurant                  Café   
32                  Hotel                Bakery           Coffee Shop   
33            Coffee Shop                 Hotel                Bakery   
62                 Bakery          Cocktail Bar                 Hotel   
82                   Café                Bakery           Coffee Shop   
97                   Café          Noodle House           Coffee Shop   
104                   Bar    Chinese Restaurant        Ice Cream Shop   
128                  Café                Bakery            Food Court   
143  Fast Food Restaurant    Chinese Restaurant   Japanese Restaurant   
210                  Park    Chinese Restaurant   Japanese Restaurant   
212                  Café    Italian Restaurant    Chinese Restaurant   
213                  Café    Italian Restaurant    Chinese Restaurant   
216            Food Court     Korean Restaurant                  Café   
231                  Café           Bus Station   Japanese Restaurant   
232                  Café           Bus Station   Japanese Restaurant   
233                  Café           Bus Station   Japanese Restaurant   
262    Chinese Restaurant   Japanese Restaurant  Fast Food Restaurant   
272        Ice Cream Shop           Coffee Shop       Thai Restaurant   
273        Ice Cream Shop           Coffee Shop       Thai Restaurant   
286                 Hotel     Korean Restaurant    Spanish Restaurant   
291    Italian Restaurant                Bakery                 Hotel   
303                  Café   Japanese Restaurant       Thai Restaurant   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1              Food Court                   Bar            Restaurant   
32             Food Court   Japanese Restaurant           Supermarket   
33             Food Court   Japanese Restaurant    Italian Restaurant   
62     Italian Restaurant                   Bar                  Café   
82    Japanese Restaurant      Ramen Restaurant            Food Court   
97                  Hotel            Food Court                Bakery   
104            Food Court         Shopping Mall                  Café   
128           Coffee Shop          Dessert Shop   Japanese Restaurant   
143        Sandwich Place            Food Court 

### Cluster 4: Popular food, hawker centers and Asian restaurants

In [371]:
tablemerged.loc[tablemerged['Cluster Labels'] == 4, tablemerged.columns[[0] + list(range(4, tablemerged.shape[1]))]]

area 1st Most Common Venue 2nd Most Common Venue  \
4    Ang Mo Kio Avenue 3            Food Court    Chinese Restaurant   
5    Ang Mo Kio Avenue 3            Food Court    Chinese Restaurant   
6    Ang Mo Kio Avenue 3            Food Court    Chinese Restaurant   
83          Geylang Road    Chinese Restaurant          Noodle House   
84          Geylang Road    Chinese Restaurant          Noodle House   
99         Highland Road    Chinese Restaurant          Noodle House   
100        Highland Road    Chinese Restaurant          Noodle House   
117      Jalan Hock Chye    Chinese Restaurant          Noodle House   
118      Jalan Hock Chye    Chinese Restaurant          Noodle House   
145          Jansen Road    Chinese Restaurant      Asian Restaurant   
146          Jansen Road    Chinese Restaurant      Asian Restaurant   
183        Lorong Sarina    Chinese Restaurant           Coffee Shop   
184        Lorong Sarina    Chinese Restaurant           Coffee Shop   
185        Lorong Sarina    Chinese Restaurant           Coffee Shop   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4             Coffee Shop  Fast Food Restaurant    Seafood Restaurant   
5             Coffee Shop  Fast Food Restaurant    Seafood Restaurant   
6             Coffee Shop  Fast Food Restaurant    Seafood Restaurant   
83             Food Court    Seafood Restaurant                 Hotel   
84             Food Court    Seafood Restaurant                 Hotel   
99             Food Court                  Café           Coffee Shop   
100            Food Court                  Café           Coffee Shop   
117           Coffee Shop            Food Court                  Café   
118           Coffee Shop            Food Court                  Café   
145           Coffee Shop          Noodle House                  Café   
146           Coffee Shop          Noodle House                  Café   
183          Noodle House            Food Court     Indian Restaurant   
184          Noodle House            Food Court     Indian Restaurant   
185          Noodle House            Food Court     Indian Restaurant   

    6th Most Common Venue          7th Most Common Venue  \
4             Supermarket  Vegetarian / Vegan Restaurant   
5             Supermarket  Vegetarian / Vegan Restaurant   
6             Supermarket  Vegetarian / Vegan Restaurant   
83     Dim Sum Restaurant               Asian Restaurant   
84     Dim Sum Restaurant               Asian Restaurant   
99       Asian Restaurant                    Bus Station   
100      Asian Restaurant                    Bus Station   
117      Asian Restaurant                    Bus Station   
118      Asian Restaurant                    Bus Station   
145            Food Court                           Park   
146            Food Court                           Park   
183    Seafood Restaurant               Asian Restaurant   
184    Seafood Restaurant               Asian Restaurant   
185    Seafood Restaurant               Asian Restaurant   

    8th Most Common Venue          9th Most Common Venue  \
4            Dessert Shop                   Noodle House   
5            Dessert Shop                   Noodle House   
6            Dessert Shop                   Noodle House   
83              BBQ Joint  Vegetarian / Vegan Restaurant   
84              BBQ Joint  Vegetarian / Vegan Restaurant   
99                   Park  Vegetarian / Vegan Restaurant   
100                  Park  Vegetarian / Vegan Restaurant   
117          Dessert Shop                         Bakery   
118          Dessert Shop                         Bakery   
145           Bus Station                    Snack Place   
146           Bus Station                    Snack Place   
183         Train Station                    Snack Place   
184         Train Station                    Snack Place   
185         Train Station                    Snack Place   

    10th Most Common Venue  
4      Japan

### Cluster 5: Indian area and restaurants

In [372]:
tablemerged.loc[tablemerged['Cluster Labels'] == 5, tablemerged.columns[[0] + list(range(4, tablemerged.shape[1]))]]

area 1st Most Common Venue 2nd Most Common Venue  \
25       Cambridge Road     Indian Restaurant    Chinese Restaurant   
52            Cuff Road     Indian Restaurant                  Café   
59        Dunlop Street     Indian Restaurant                  Café   
111         Jalan Besar     Indian Restaurant    Chinese Restaurant   
235    Race Course Road     Indian Restaurant    Chinese Restaurant   
249      Serangoon Road     Indian Restaurant    Chinese Restaurant   
299  Upper Dickson Road     Indian Restaurant                  Café   

    3rd Most Common Venue 4th Most Common Venue  \
25                   Café                 Hotel   
52     Chinese Restaurant                Bakery   
59     Chinese Restaurant                 Hotel   
111                  Café                Hostel   
235                  Café                 Hotel   
249                  Café                 Hotel   
299    Chinese Restaurant                 Hotel   

             5th Most Common Venue          6th Most Common Venue  \
25                     Coffee Shop                     Food Court   
52                           Hotel  Vegetarian / Vegan Restaurant   
59   Vegetarian / Vegan Restaurant                         Bakery   
111                      BBQ Joint                          Hotel   
235                   Dessert Shop                         Hostel   
249  Vegetarian / Vegan Restaurant                         Hostel   
299  Vegetarian / Vegan Restaurant                         Bakery   

    7th Most Common Venue          8th Most Common Venue  \
25             Restaurant             Seafood Restaurant   
52       Asian Restaurant                   Dessert Shop   
59             Restaurant               Asian Restaurant   
111            Restaurant  Vegetarian / Vegan Restaurant   
235            Food Court                   Noodle House   
249             BBQ Joint               Asian Restaurant   
299            Restaurant               Asian Restaurant   

    9th Most Common Venue         10th Most Common Venue  
25       Asian Restaurant            Japanese Restaurant  
52     Dim Sum Restaurant                      BBQ Joint  
59                 Hostel                   Dessert Shop  
111                Bakery                    Supermarket  
235            Soup Place  Vegetarian / Vegan Restaurant  
249    Dim Sum Restaurant                   Dessert Shop  
299                Hostel                   Dessert Shop

### Cluster 6: Entertainment, leisure, parks, utilities and food

In [373]:
tablemerged.loc[tablemerged['Cluster Labels'] == 6, tablemerged.columns[[0] + list(range(4, tablemerged.shape[1]))]]

area 1st Most Common Venue  \
0                   Almond Street                  Park   
2             Anchorvale Crescent  Fast Food Restaurant   
8                   Ashwood Grove  Fast Food Restaurant   
10                   Bedok Avenue                 Beach   
13                 Bedok Ria Walk          Noodle House   
14                     Bedok Road      Asian Restaurant   
15                Belgravia Drive                  Café   
16                Belgravia Drive                  Café   
17                     Bilal Lane              Bus Line   
21                Brookvale Drive                  Café   
28                 Cashew Terrace  Fast Food Restaurant   
37                   Chin Terrace              Bus Stop   
38        Chinese Cemetery Path 4             Gun Range   
40         Choa Chu Kang Avenue 5           Supermarket   
41             Choa Chu Kang Loop  Fast Food Restaurant   
42          Choa Chu Kang North 6  Fast Food Restaurant   
46              Compassvale Drive  Fast Food Restaurant   
47                   Coral Island                 Plaza   
49               Cornwall Gardens                Bakery   
51               Corporation Walk            Food Court   
55            Dairy Farm Crescent                  Café   
56                      Deal Road           Coffee Shop   
57                     Dedap Link   American Restaurant   
61                 Dunsfold Drive    Seafood Restaurant   
65                 Eastwood Drive           Golf Course   
66               Eastwood Terrace           Golf Course   
68                Eng Kong Garden                  Café   
72                  Faber Terrace           Bus Station   
76                  Florissa Park           Bus Station   
77                  Florissa Park           Bus Station   
80                    Gangsa Road            Food Court   
89                Greenleaf Drive           Bus Station   
90               Greenmead Avenue    Italian Restaurant   
94                      Gul Drive            Smoke Shop   
98             Hazel Park Terrace           Coffee Shop   
102                 Holland Green           Bus Station   
105                 Hong San Walk  Fast Food Restaurant   
112                   Jalan Besut            Food Court   
115                Jalan Gumilang           Coffee Shop   
116              Jalan Haji Salam           Bus Station   
119         Jalan Kampong Chantek                   Gym   
126            Jalan Loyang Besar                  Park   
127                  Jalan Mariam           Bus Station   
131                 Jalan Pelatok           Bus Station   
132           Jalan Pokok Serunai    Chinese Restaurant   
138                 Jalan Siantan           Bus Station   
141                  Jalan Telang    Chinese Restaurant   
151           Kaki Bukit Avenue 1           Coffee Shop   
153                Kay Siang Road    Italian Restaurant   
156                  Kew Crescent           Coffee Shop   
157                   King'S Walk                Bakery   
162                  Lentor Green           Bus Station   
163                   Lentor Walk           Bus Station   
165           Lim Chu Kang Lane 3                  Farm   
166           Lim Chu Kang Lane 5                  Farm   
167                  Limau Garden           Coffee Shop   
171             Lorong 106 Changi    Chinese Restaurant   
187                   Loyang Rise           Bus Station   
188                   Loyang Rise           Bus Station   
189               Luxus Hill View            Food Court   
192               Marsiling Drive            Food Court   
196                   Meragi Road  Fast Food Restaurant   
197                   Merryn Road           Bus Station   
198                Miltonia Close           Coffee Shop   
199                   Mimosa Road          Dessert Shop   
200                  Minaret Walk      Malay Restaurant   
201                Ming Teck Park           Bus Station   
202            Mo

### Cluster 7: Upscale venues

In [374]:
tablemerged.loc[tablemerged['Cluster Labels'] == 7, tablemerged.columns[[0] + list(range(4, tablemerged.shape[1]))]]

area 1st Most Common Venue 2nd Most Common Venue  \
24      Cairnhill Road                 Hotel              Boutique   
67   Emerald Hill Road   Japanese Restaurant                 Hotel   
114         Jalan Elok              Boutique                 Hotel   
164        Leonie Hill              Boutique         Shopping Mall   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
24    Japanese Restaurant    Chinese Restaurant                Bakery   
67               Boutique         Shopping Mall        Clothing Store   
114   Japanese Restaurant      Sushi Restaurant                Bakery   
164   Japanese Restaurant       Bubble Tea Shop                 Hotel   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
24       Sushi Restaurant         Shopping Mall       Bubble Tea Shop   
67       Sushi Restaurant    Chinese Restaurant              Wine Bar   
114       Bubble Tea Shop    Chinese Restaurant         Shopping Mall   
164    Chinese Restaurant           Coffee Shop        Cosmetics Shop   

    9th Most Common Venue 10th Most Common Venue  
24         Cosmetics Shop            Supermarket  
67        Bubble Tea Shop      Korean Restaurant  
114        Cosmetics Shop               Tea Room  
164      Sushi Restaurant         Clothing Store

### Cluster 8: Coffee shops and traditional restaurants in residential areas

In [375]:
tablemerged.loc[tablemerged['Cluster Labels'] == 8, tablemerged.columns[[0] + list(range(4, tablemerged.shape[1]))]]

area 1st Most Common Venue 2nd Most Common Venue  \
31             Cavan Road                  Café    Chinese Restaurant   
58            Dunbar Walk           Coffee Shop    Chinese Restaurant   
75         Fidelio Street           Coffee Shop     Indian Restaurant   
78          Fourth Street           Coffee Shop     Indian Restaurant   
79         Frankel Street    Chinese Restaurant           Coffee Shop   
121           Jalan Keris           Coffee Shop          Noodle House   
122           Jalan Keris           Coffee Shop          Noodle House   
142    Jalan Terang Bulan           Coffee Shop    Chinese Restaurant   
144           Jalan Yasin    Chinese Restaurant           Coffee Shop   
160       Lavender Street    Chinese Restaurant                  Café   
169  Lorong 1 Realty Park           Coffee Shop    Chinese Restaurant   
175  Lorong 7 Realty Park           Coffee Shop    Chinese Restaurant   
176      Lorong Abu Talib           Coffee Shop    Chinese Restaurant   
179  Lorong M Telok Kurau           Coffee Shop    Chinese Restaurant   
205           Nallur Road    Chinese Restaurant    Seafood Restaurant   
251        Siglap Gardens    Seafood Restaurant           Coffee Shop   
253           Siglap Road           Coffee Shop     Indian Restaurant   
254           Siglap Road           Coffee Shop     Indian Restaurant   
269      Swan Lake Avenue           Coffee Shop    Chinese Restaurant   
311        Wilton Gardens           Coffee Shop    Chinese Restaurant   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
31            Coffee Shop     Indian Restaurant                Hostel   
58      Indian Restaurant          Noodle House      Asian Restaurant   
75     Chinese Restaurant           Supermarket          Dessert Shop   
78     Chinese Restaurant                  Café          Dessert Shop   
79      Indian Restaurant          Noodle House      Asian Restaurant   
121                  Café    Chinese Restaurant          Dessert Shop   
122                  Café    Chinese Restaurant          Dessert Shop   
142                  Café      Asian Restaurant           Supermarket   
144     Indian Restaurant            Food Court          Noodle House   
160                 Hotel     Indian Restaurant           Coffee Shop   
169            Food Court                  Café      Asian Restaurant   
175            Food Court                  Café  Fast Food Restaurant   
176     Indian Restaurant                  Café          Dessert Shop   
179          Noodle House      Asian Restaurant     Indian Restaurant   
205     Indian Restaurant      Asian Restaurant           Bus Station   
251    Chinese Restaurant     Indian Restaurant                  Café   
253    Chinese Restaurant    Seafood Restaurant                  Café   
254    Chinese Restaurant    Seafood Restaurant                  Café   
269           Supermarket            Food Court          Dessert Shop   
311                  Café     Indian Restaurant          Dessert Shop   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
31                  Hotel             BBQ Joint          Noodle House   
58             Food Court             Gastropub    Italian Restaurant   
75                   Café      Asian Restaurant           Bus Station   
78              Gastropub           Pizza Place    Seafood Restaurant   
79             Food Court    Seafood Restaurant                  Café   
121           Bus Station                  Park      Asian Restaurant   
122           Bus Station                  Park      Asian Restaurant   
142            Food Court           Bus Station                Bakery   
144    Seafood Restaurant      Asian Restaurant             BBQ Joint   
160          Noodle House             BBQ Joint                Hostel   
169  Fast Food Restaurant           Snack Place          Dessert Shop   
175      Asian Restaurant         Shopping Mall                Bakery   
176 

### Cluster 9: Food around Serangoon

In [376]:
tablemerged.loc[tablemerged['Cluster Labels'] == 9, tablemerged.columns[[0] + list(range(4, tablemerged.shape[1]))]]

area 1st Most Common Venue 2nd Most Common Venue  \
3         Ang Mo Kio Avenue 2    Chinese Restaurant            Food Court   
18            Blandford Drive    Chinese Restaurant            Food Court   
19               Bodmin Drive    Chinese Restaurant            Food Court   
22        Bukit Merah Central    Chinese Restaurant           Coffee Shop   
23             Burghley Drive    Chinese Restaurant            Food Court   
27              Cardiff Grove    Chinese Restaurant   Japanese Restaurant   
29             Casuarina Road           Bus Station    Chinese Restaurant   
30             Casuarina Road           Bus Station    Chinese Restaurant   
36        Cheng Soon Crescent     Korean Restaurant    Chinese Restaurant   
39          Chiselhurst Grove    Chinese Restaurant                  Park   
48                Corfe Place    Chinese Restaurant            Food Court   
54              Da Silva Lane    Chinese Restaurant          Noodle House   
88                 Grace Walk    Chinese Restaurant   Japanese Restaurant   
101             Hindhede Road     Korean Restaurant    Chinese Restaurant   
107                Hythe Road    Chinese Restaurant            Food Court   
110           Jalan Belangkas            Food Court      Asian Restaurant   
125              Jalan Limbok    Chinese Restaurant            Food Court   
134             Jalan Rengkam    Chinese Restaurant          Noodle House   
137               Jalan Setia    Chinese Restaurant            Food Court   
152        Kang Choo Bin Walk    Chinese Restaurant           Coffee Shop   
154           Kee Choe Avenue            Food Court    Chinese Restaurant   
155           Keppel Bay View    Chinese Restaurant            Food Court   
158                Kovan Road          Noodle House           Coffee Shop   
181        Lorong Pisang Emas    Chinese Restaurant     Korean Restaurant   
182        Lorong Pisang Raja    Chinese Restaurant     Korean Restaurant   
190           Macpherson Road            Food Court    Chinese Restaurant   
193               Mattar Road            Food Court    Chinese Restaurant   
194          Mayflower Avenue            Food Court    Chinese Restaurant   
195            Mayflower Rise            Food Court    Chinese Restaurant   
228        Portchester Avenue            Food Court    Chinese Restaurant   
247  Serangoon North Avenue 1    Chinese Restaurant           Coffee Shop   
255                Simon Lane    Chinese Restaurant          Noodle House   
281              Tannery Lane    Chinese Restaurant            Food Court   
290              Toh Yi Drive     Korean Restaurant    Chinese Restaurant   
300        Upper Thomson Road    Chinese Restaurant                  Café   
304            Wan Tho Avenue            Food Court    Chinese Restaurant   
319                 York Road    Chinese Restaurant            Food Court   

             3rd Most Common Venue 4th Most Common Venue  \
3                 Asian Restaurant             Pet Store   
18                     Bus Station                  Park   
19                            Park                Bakery   
22                      Food Court           Bus Station   
23                     Bus Station                  Park   
27                     Coffee Shop           Bus Station   
29                      Food Court            Restaurant   
30                      Food Court            Restaurant   
36                            Café            Food Court   
39                     Coffee Shop                   Pub   
48                          Bakery                  Café   
54                      Food Court           Coffee Shop   
88                     Coffee Shop          Noodle House   
101                           Café                 Trail   
107               Asian Restaurant           Coffee Shop   
110             Chinese Restaurant           Coffee Shop   
125                    Coffee Shop      Asian Restaurant   
134               

In [377]:
tablemerged

area  latitude  longitude  Cluster Labels  \
0                   Almond Street    1.3725   103.7724               6   
1                     Amoy Street    1.2801   103.8468               3   
2             Anchorvale Crescent    1.3985   103.8921               6   
3             Ang Mo Kio Avenue 2    1.3774   103.8350               9   
4             Ang Mo Kio Avenue 3    1.3676   103.8446               4   
5             Ang Mo Kio Avenue 3    1.3708   103.8435               4   
6             Ang Mo Kio Avenue 3    1.3699   103.8421               4   
7                       Ash Grove    1.3329   103.8030               2   
8                   Ashwood Grove    1.4288   103.7767               6   
9                     Barker Road    1.3214   103.8351               2   
10                   Bedok Avenue    1.3210   103.9530               6   
11           Bedok Reservoir Road    1.3349   103.9150               0   
12           Bedok Reservoir Road    1.3360   103.9249               0   
13                 Bedok Ria Walk    1.3260   103.9535               6   
14                     Bedok Road    1.3307   103.9484               6   
15                Belgravia Drive    1.3804   103.8653               6   
16                Belgravia Drive    1.3802   103.8660               6   
17                     Bilal Lane    1.3225   103.9530               6   
18                Blandford Drive    1.3600   103.8615               9   
19                   Bodmin Drive    1.3615   103.8625               9   
20                 Boon Lay Drive    1.3418   103.7119               0   
21                Brookvale Drive    1.3300   103.7683               6   
22            Bukit Merah Central    1.2821   103.8181               9   
23                 Burghley Drive    1.3608   103.8599               9   
24                 Cairnhill Road    1.3077   103.8372               7   
25                 Cambridge Road    1.3139   103.8471               5   
26                 Canberra Drive    1.4380   103.8310               0   
27                  Cardiff Grove    1.3572   103.8661               9   
28                 Cashew Terrace    1.3719   103.7667               6   
29                 Casuarina Road    1.3756   103.8282               9   
30                 Casuarina Road    1.3749   103.8282               9   
31                     Cavan Road    1.3113   103.8617               8   
32             Chancery Hill Walk    1.3232   103.8399               3   
33                  Chancery Lane    1.3219   103.8400               3   
34                    Changi Road    1.3194   103.9131               0   
35                  Charlton Road    1.3541   103.8819               1   
36            Cheng Soon Crescent    1.3368   103.7735               9   
37                   Chin Terrace    1.3646   103.9739               6   
38        Chinese Cemetery Path 4    1.3736   103.6872               6   
39              Chiselhurst Grove    1.3591   103.8657               9   
40         Choa Chu Kang Avenue 5    1.3744   103.7396               6   
41             Choa Chu Kang Loop    1.3844   103.7461               6   
42          Choa Chu Kang North 6    1.3963   103.7437               6   
43                     Clover Way    1.3496   103.8442               0   
44                     Cluny Park    1.3172   103.8145               2   
45                     Cluny Road    1.3175   103.8169               2   
46              Compassvale Drive    1.3902   103.8946               6   
47                   Coral Island    1.2514   103.8455               6   
48                    Corfe Place    1.3600   103.8606               9   
49               Cornwall Gardens    1.3183   103.8019               6   
50                Coronation Road    1.3220   103.8093               2   
51               Corporation Walk    1.3396   103.7182               6   
52                      Cuff Road    1.3071   103.8536               5   
53                 Cypress Avenue

## Plotting recommended cluster

In [392]:
folium.Circle(location=[1.3282, 103.8083], radius=2000, 
        popup='Laurelhurst Park', line_color='#3186cc',
        fill_color='#3186cc', fill=True).add_to(map_clusters)

In [393]:
map_clusters